train.py

```diff
@@ -312,47 +315,46 @@ def log_step(step_type, device, step, log_output=None, tracker=None):
-        task = tasks.setup_task(args)
-
-        # Load valid dataset (we load training data below, based on the latest checkpoint)
-        for valid_sub_split in args.valid_subset.split(','):
-            task.load_dataset(valid_sub_split, combine=True, epoch=0)
-
-        # Build models and criteria to print some metadata
-        model_parallel = dp.DataParallel(
-            lambda: task.build_model(args), device_ids=devices)
-        model, criterion = task.build_model(args), task.build_criterion(args)
-        print_model_criterion(model, criterion, args)
-        del model, criterion
-
-        # Build trainers
-        trainers = {
-            device: Trainer(args, task, model, task.build_criterion(args), xla=True)
-            for device, model in zip(model_parallel.devices, model_parallel.models)
-        }
-        lr = trainers[devices[0]].get_lr()
-
-        # Load the latest checkpoint if one is available and restore the
-        # corresponding train iterator
-        extra_state, epoch_itr = checkpoint_utils.load_checkpoint(
-            args, trainers[devices[0]])
-        if extra_state is not None:
-            # checkpoint detected, load saved model weights to all devices
-            xu.eprint(
-                'checkpoint detected, device 0 meters need to be '
-                're-loaded to device'
-            )
-            checkpoint_utils.load_checkpoint_tpu(args, trainers, devices[0])
-        valid_subsets = args.valid_subset.split(',')
-        return task, trainers, model_parallel, epoch_itr, lr, valid_subsets
-
-    def train_loop_fn(model, loader, device, context):
-        trainer = trainers[str(device)]
-        stats, log_output = None, None
-        tracker = xm.RateTracker()
+    def prepare_task(args, xla_device):
+       # Setup task, e.g., translation, language modeling, etc.
+       task = tasks.setup_task(args)
+
+       # Load valid dataset (we load training data below, based on the latest checkpoint)
+       for valid_sub_split in args.valid_subset.split(','):
+           task.load_dataset(valid_sub_split, combine=True, epoch=0)
+
+       # Build models and criteria to print some metadata
+       torch.manual_seed(args.seed)
+       model, criterion = task.build_model(args), task.build_criterion(args)
+       xm.master_print(model)
+       xm.master_print('| model {}, criterion {}'.format(
+           args.arch, criterion.__class__.__name__))
+       xm.master_print('| num. model params: {} (num. trained: {})'.format(
+           sum(p.numel() for p in model.parameters()),
+           sum(p.numel() for p in model.parameters() if p.requires_grad)))
+       model = model.to(xla_device)
+       trainer = Trainer(args, task, model, criterion, xla=True)
+       lr = trainer.get_lr()
+
+       # Load the latest checkpoint if one is available and restore the
+       # corresponding train iterator
+       # set distributed args here to shard data
+       trainer.args.distributed_rank = xm.get_ordinal()
```

Comment:
Why are we doing this if we override below in lines 345-346? Is it just required to load checkpoint?

Reply:
yes, it doesn't work otherwise.